src/structarray.jl

```diff
@@ -463,9 +479,20 @@ Base.@pure cst(::Type{SA}) where {SA} = combine_style_types(array_types(SA).para
 
 BroadcastStyle(::Type{SA}) where {SA<:StructArray} = StructArrayStyle{typeof(cst(SA)), ndims(SA)}()
 
-function Base.similar(bc::Broadcasted{StructArrayStyle{S, N}}, ::Type{ElType}) where {S<:DefaultArrayStyle, N, ElType}
-    ContainerType = isnonemptystructtype(ElType) ? StructArray{ElType} : Array{ElType}
-    return similar(ContainerType, axes(bc))
+# Here we use `similar` defined for `S` to build the dest Array.
+function Base.similar(bc::Broadcasted{StructArrayStyle{S, N}}, ::Type{ElType}) where {S, N, ElType}
+    bc′ = convert(Broadcasted{S}, bc)
```

Comment:
Could this allocate? My concern is that making a copy here would be unnecessary, as we are just using this object to call `similar`.

Reply:
Our `Base.broadcast` use `convert(Broadcasted{Nothing}, bc)` to activate the default `copyto!`.
So I guess it's free cost at run-time?

Reply:
Ah, you're right, looks like it's just changing a type parameter [here](https://github.com/JuliaLang/julia/blob/ac922b0ad4295e13ec553abcab3b954e4a60064f/base/broadcast.jl#L197).